File: src/providers/gemini.js
import { parseJsonFromText } from "../core/json.js";

export class GeminiProvider {
  constructor(options = {}) {
    this.apiKey = options.apiKey ?? process.env.GEMINI_API_KEY;
    this.model = options.model ?? process.env.GEMINI_MODEL ?? "gemini-2.5-flash";
    this.baseUrl = options.baseUrl ?? "https://generativelanguage.googleapis.com/v1beta";
  }

  get available() {
    return Boolean(this.apiKey);
  }

  async generate({ system, prompt, json = false, temperature = 0.2, googleSearch = false }) {
    if (!this.apiKey) {
      throw new Error("GEMINI_API_KEY is required unless offline mode is enabled");
    }

    const url = `${this.baseUrl}/models/${this.model}:generateContent?key=${encodeURIComponent(this.apiKey)}`;
    const body = {
      systemInstruction: system
        ? {
            parts: [{ text: system }]
          }
        : undefined,
      contents: [
        {
          role: "user",
          parts: [{ text: prompt }]
        }
      ],
      generationConfig: {
        temperature,
        ...(json && !googleSearch ? { responseMimeType: "application/json" } : {})
      },
      ...(googleSearch ? { tools: [{ googleSearch: {} }] } : {})
    };

    const response = await fetch(url, {
      method: "POST",
      headers: {
        "content-type": "application/json"
      },
      body: JSON.stringify(body)
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Gemini ${this.model} request failed (${response.status}): ${errorText}`);
    }

    const payload = await response.json();
    const text = payload.candidates?.[0]?.content?.parts
      ?.map((part) => part.text ?? "")
      .join("")
      .trim();

    if (!text) {
      throw new Error(`Gemini ${this.model} returned no text`);
    }

    return text;
  }

  async generateJson(args) {
    const text = await this.generate({ ...args, json: true });
    return parseJsonFromText(text);
  }
}

export function createModel({ offline = false, logger } = {}) {
  const provider = new GeminiProvider();
  if (!provider.available || offline) {
    if (!provider.available) {
      logger?.warn("GEMINI_API_KEY not set; using deterministic offline agents.");
    }
    return null;
  }
  return provider;
}

File: go_issue_agent/providers/gemini.py
from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request

from go_issue_agent.core.json_utils import parse_json_from_text


class GeminiProvider:
    def __init__(self, api_key: str | None = None, model: str | None = None, base_url: str | None = None):
        self.api_key = api_key or os.environ.get("GEMINI_API_KEY")
        self.model = model or os.environ.get("GEMINI_MODEL", "gemini-2.5-flash")
        self.base_url = base_url or "https://generativelanguage.googleapis.com/v1beta"

    @property
    def available(self) -> bool:
        return bool(self.api_key)

    def generate(self, system: str | None, prompt: str, json_mode: bool = False, temperature: float = 0.2, google_search: bool = False) -> str:
        if not self.api_key:
            raise RuntimeError("GEMINI_API_KEY is required unless offline mode is enabled")

        url = f"{self.base_url}/models/{self.model}:generateContent?key={urllib.parse.quote(self.api_key)}"
        body = {
            "systemInstruction": {"parts": [{"text": system}]} if system else None,
            "contents": [{"role": "user", "parts": [{"text": prompt}]}],
            "generationConfig": {"temperature": temperature},
        }
        if json_mode and not google_search:
            body["generationConfig"]["responseMimeType"] = "application/json"
        if google_search:
            body["tools"] = [{"googleSearch": {}}]

        request = urllib.request.Request(
            url,
            data=json.dumps({key: value for key, value in body.items() if value is not None}).encode("utf-8"),
            headers={"content-type": "application/json"},
            method="POST",
        )
        try:
            with urllib.request.urlopen(request) as response:
                payload = json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as error:
            detail = error.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"Gemini {self.model} request failed ({error.code}): {detail}") from error

        parts = payload.get("candidates", [{}])[0].get("content", {}).get("parts", [])
        text = "".join(part.get("text", "") for part in parts).strip()
        if not text:
            raise RuntimeError(f"Gemini {self.model} returned no text")
        return text

    def generate_json(self, **kwargs):
        return parse_json_from_text(self.generate(json_mode=True, **kwargs))


def create_model(offline: bool = False, logger=None):
    provider = GeminiProvider()
    if offline or not provider.available:
        if not provider.available and logger:
            logger.warn("GEMINI_API_KEY not set; using deterministic offline agents.")
        return None
    return provider
